packages/blade/src/components/Tabs/utils.ts

```diff
@@ -0,0 +1,85 @@
+import React from 'react';
+import type { TabsProps } from './types';
+import { getComponentId } from '~utils/isValidAllowedChildren';
+import { logger, throwBladeError } from '~utils/logger';
+import { useIsomorphicLayoutEffect } from '~utils/useIsomorphicLayoutEffect';
+
+const iconSizeMap = {
+  medium: 'medium',
+  large: 'large',
+} as const;
+
+const badgeSizeMap = {
+  medium: 'small',
+  large: 'medium',
+} as const;
+
+const counterSizeMap = {
+  medium: 'small',
+  large: 'small',
+} as const;
+
+const propRestrictionMap = {
+  Badge: {
+    medium: {
+      size: badgeSizeMap.medium,
+    },
+    large: {
+      size: badgeSizeMap.large,
+    },
+  },
+  Counter: {
+    medium: {
+      size: counterSizeMap.medium,
+    },
+    large: {
+      size: counterSizeMap.large,
+    },
+  },
+} as const;
+
+type TrailingComponents = keyof typeof propRestrictionMap;
+
+const useTabsItemPropRestriction = (
+  trailing: React.ReactNode,
+  tabItemSize: NonNullable<TabsProps['size']>,
+): React.ReactNode => {
+  const [
+    validatedTrailingComponent,
+    setValidatedTrailingComponent,
+  ] = React.useState<React.ReactElement | null>(null);
+
+  // validate and restrict sub component props in trailing prop
+  useIsomorphicLayoutEffect(() => {
+    if (React.isValidElement(trailing)) {
+      const trailingComponentType = getComponentId(trailing) as TrailingComponents;
+      const restrictedProps = propRestrictionMap[trailingComponentType]?.[tabItemSize];
+      if (__DEV__) {
+        if (!restrictedProps) {
+          throwBladeError({
+            message: `Only Badge or Counter component is accepted as trailing`,
+            moduleName: 'TabsItem',
+          });
+        }
+
+        const restrictedPropKeys = Object.keys(restrictedProps);
+        for (const prop of restrictedPropKeys) {
+          if (trailing?.props?.hasOwnProperty(prop)) {
+            logger({
+              message: `Do not pass "${prop}" to "${trailingComponentType}" while inside TabsItem trailing, because we override it.`,
+              moduleName: 'TabsItem',
+              type: 'warn',
```

Comment:
shouldn't this be error instead of warning? we shall break the component rendering IMO in these kind of cases

Reply:
Let's say internally we override `size` prop of Badge to always be `large`. 
And consumer again passes `size=medium` but this won't have an effect because we override it internally. 

In this case even if consumer passes a size prop nothing breaks visually, it's essentially becomes a NOOP. 

Throwing error for each and every case will cause disruption while using this component instead it's better to let them know this has no effect even if you do it. (it's even worse on RN because RN will show a full page error dialog)

-----

If passing some prop can cause visual breakage in the component only then we should opt for throwing errors, like how we throw error if they pass anything other than Badge/Counter.